src/rp2_common/pico_standard_link/crt0.S

```diff
@@ -144,39 +144,53 @@ __unhandled_user_irq:
 unhandled_user_irq_num_in_r0:
     bkpt #0
 
-.section .reset, "ax"
-
-// This is the beginning of the image, which is entered from stage2 or bootrom USB MSD watchdog reboot
+// ----------------------------------------------------------------------------
 
-// note if we are NO_FLASH then start: below is currently identical anyway, so save 4 bytes
-#if !PICO_NO_FLASH
-    // We simply install our own vector table and redirect through it
-    ldr r0, =__vectors
-    b __vector_entry
-#endif
+.section .reset, "ax"
 
-// ELF entry point generally called when we load an ELF via debugger
+// On flash builds, the vector table comes first in the image (conventional).
+// On NO_FLASH builds, the reset handler section comes first, as the entry
+// point is at offset 0 (fixed due to bootrom), and VTOR is highly-aligned.
+// Image is entered in various ways:
+//
+// - NO_FLASH builds are entered from beginning by UF2 bootloader
+//
+// - Flash builds vector through the table into _reset_handler from boot2
+//
+// - Either type can be entered via _entry_point by the debugger, and flash builds
+//   must then be sent back round the boot sequence to properly initialise flash
+
+// ELF entry point:
 .type _entry_point,%function
 .thumb_func
 .global _entry_point
 _entry_point:
 
 #if PICO_NO_FLASH
-    // non flash
+    // Vector through our own table (SP, VTOR will not have been set up at
+    // this point). Same path for debugger entry and bootloader entry.
     ldr r0, =__vectors
 #else
-    // todo clear watchdog?
-    // When using flash, we install and use the ROM vector table to go thru regular bootrom/stage2 flash sequence
+    // Debugger tried to run code after loading, so SSI is in 03h-only mode.
+    // Go back through bootrom + boot2 to properly initialise flash.
     movs r0, #0
 #endif
-
-__vector_entry:
-    ldr r1, =(PPB_BASE + M0PLUS_CPUID_OFFSET)
-    str r0, [r1, #8]
+    ldr r1, =(PPB_BASE + M0PLUS_VTOR_OFFSET)
```

Comment:
Thank you!!

Reply:
Yeah I had already fixed in passing, but I laughed when you mentioned it in the stream.

Reply:
Any chance you can not have both the processor model identification register and the core number identification register not both called CPUID whilst you're at it? :)

Reply:
> the processor model identification register

That's `M0PLUS: CPUID Register` on page 80 of the RP2040 datasheet. 
https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2040/hardware_regs/include/hardware/regs/m0plus.h#L565
https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2040/hardware_structs/include/hardware/structs/scb.h#L14

> the core number identification register

That's `SIO: CPUID Register` on page 42 of the RP2040 datasheet.
https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2040/hardware_regs/include/hardware/regs/sio.h#L21
https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2040/hardware_structs/include/hardware/structs/sio.h#L15

> not both called CPUID whilst you're at it? :)

I suspect it's too late to change that at this stage. And I'm sure there's probably other registers which also have "duplicated names" between different peripheral blocks?

Reply:
It's a bit late in the day to rename registers, but there is at least a documentation issue there, so I'll drop a couple of NOTE boxes in the relevant parts of the datasheet. I think you were lead down the garden path a little by this literal value appearing in the crt0 disassembly.

Reply:
Note that the core number check in crt0 should be redundant anyway. It should be impossible to reach this point on core 1 (see https://github.com/raspberrypi/pico-bootrom/blob/00a4a19114195e20fb817bdfbca1165e157eef37/bootrom/bootrom_rt0.S#L222), there's just a paranoia check there in case you do something spectacularly broken